File: dotnet/samples/03-workflows/agents-in-workflows/AgentsInWorkflows.cs
// Copyright (c) Microsoft. All rights reserved.
// Description: Using AI agents as workflow steps in a translation chain.
// Docs: https://learn.microsoft.com/agent-framework/workflows/overview

using Azure.AI.OpenAI;
using Azure.Identity;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;

namespace WorkflowSamples.AgentsInWorkflows;

// <agents_in_workflows>
/// <summary>
/// Demonstrates using AI agents as executors within a workflow.
/// Three translation agents are connected sequentially:
/// French Agent → Spanish Agent → English Agent
/// </summary>
public static class Program
{
    private static async Task Main()
    {
        // Set up the Azure OpenAI client
        var endpoint = Environment.GetEnvironmentVariable("AZURE_OPENAI_ENDPOINT") ?? throw new InvalidOperationException("AZURE_OPENAI_ENDPOINT is not set.");
        var deploymentName = Environment.GetEnvironmentVariable("AZURE_OPENAI_DEPLOYMENT_NAME") ?? "gpt-4o-mini";
        var chatClient = new AzureOpenAIClient(new Uri(endpoint), new AzureCliCredential()).GetChatClient(deploymentName).AsIChatClient();

        // Create translation agents
        AIAgent frenchAgent = GetTranslationAgent("French", chatClient);
        AIAgent spanishAgent = GetTranslationAgent("Spanish", chatClient);
        AIAgent englishAgent = GetTranslationAgent("English", chatClient);

        // Build the workflow by chaining agents sequentially
        var workflow = new WorkflowBuilder(frenchAgent)
            .AddEdge(frenchAgent, spanishAgent)
            .AddEdge(spanishAgent, englishAgent)
            .Build();

        // Execute the workflow
        await using StreamingRun run = await InProcessExecution.StreamAsync(workflow, new ChatMessage(ChatRole.User, "Hello World!"));

        // Send the turn token to trigger the agents
        await run.TrySendMessageAsync(new TurnToken(emitEvents: true));
        await foreach (WorkflowEvent evt in run.WatchStreamAsync())
        {
            if (evt is AgentResponseUpdateEvent executorComplete)
            {
                Console.WriteLine($"{executorComplete.ExecutorId}: {executorComplete.Data}");
            }
        }
    }

    private static ChatClientAgent GetTranslationAgent(string targetLanguage, IChatClient chatClient) =>
        new(chatClient, $"You are a translation assistant that translates the provided text to {targetLanguage}.");
}
// </agents_in_workflows>

File: python/samples/03-workflows/agents-in-workflows/agents_in_workflows.py
# Copyright (c) Microsoft. All rights reserved.

"""
Agents in Workflows Sample

Demonstrates using AI agents directly as workflow steps. A Writer agent creates
content, and a Reviewer agent evaluates and provides feedback.

What you'll learn:
- Creating agents from AzureOpenAIChatClient
- Using agents as workflow executors
- Building a two-agent sequential workflow

Related samples:
- ../sequential/ — Basic sequential workflows without agents
- ../concurrent/ — Fan-out/fan-in with multiple agents

Docs: https://learn.microsoft.com/agent-framework/workflows/overview
"""

import asyncio
from typing import cast

from agent_framework import AgentResponse, WorkflowBuilder
from agent_framework.azure import AzureOpenAIChatClient
from azure.identity import AzureCliCredential


# <workflow_definition>
async def main():
    """Build and run a simple two-node agent workflow: Writer then Reviewer."""
    chat_client = AzureOpenAIChatClient(credential=AzureCliCredential())

    writer_agent = chat_client.as_agent(
        instructions=(
            "You are an excellent content writer. You create new content and edit contents based on the feedback."
        ),
        name="writer",
    )

    reviewer_agent = chat_client.as_agent(
        instructions=(
            "You are an excellent content reviewer."
            "Provide actionable feedback to the writer about the provided content."
            "Provide the feedback in the most concise manner possible."
        ),
        name="reviewer",
    )

    workflow = WorkflowBuilder(start_executor=writer_agent).add_edge(writer_agent, reviewer_agent).build()
# </workflow_definition>

    # <running>
    events = await workflow.run("Create a slogan for a new electric SUV that is affordable and fun to drive.")

    outputs = cast(list[AgentResponse], events.get_outputs())
    for output in outputs:
        print(f"{output.messages[0].author_name}: {output.text}\n")

    print("Final state:", events.get_final_state())
    # </running>

    """
    Sample Output:

    writer: "Charge Ahead: Affordable Adventure Awaits!"

    reviewer: - Consider emphasizing both affordability and fun in a more dynamic way.
    - Try using a catchy phrase that includes a play on words.
    - Ensure the slogan is succinct while capturing the essence of the car's unique selling proposition.

    Final state: WorkflowRunState.IDLE
    """


if __name__ == "__main__":
    asyncio.run(main())
